File: 81-search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.ts
function search(nums: number[], target: number): boolean {
        let left: number = 0
        let right: number = nums.length - 1;

    while (left < right) {
        let mid = Math.floor((right + left) / 2); // lower mid
        if (nums[mid] === target) return true
        if (nums[mid] === nums[left]) left++
        if (nums[mid] == nums[right]) right--; // the only difference between Q.33
        else if (nums[mid] < nums[right]) { // right part sorted
            if (target > nums[mid] && target <= nums[right]) left = mid + 1;
            else right = mid
        } else{
            if (target > nums[mid] || target < nums[left]) left = mid + 1;
            else right = mid;
        } 
    }
    return nums[left]==target;
};

File: 81-search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        l, r = 0, len(nums) - 1

        while l <= r:
            m = (l + r) // 2
            if nums[m] == target:
                return True
            if nums[l] < nums[m]:
                if nums[l] <= target < nums[m]:
                    r = m - 1
                else:
                    l = m + 1
            elif nums[l] > nums[m]:
                if nums[m] < target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1
            else:
                l += 1
        return False    
"""
[3,3,3,1,2,3,3]
"""